classic/interfaces.go

```diff
@@ -0,0 +1,367 @@
+package main
+
+import (
+	"math/big"
+
+	"github.com/openrelayxyz/plugeth-utils/core"
+	"github.com/openrelayxyz/plugeth-utils/restricted/params"
+	"github.com/openrelayxyz/plugeth-utils/restricted/types"
+)
+
+type cacheOrDataset interface {
+	*cache | *dataset
+}
+
+type ChainHeaderReader interface {
+	// Config retrieves the blockchain's chain configuration.
+	Config() *params.ChainConfig
+
+	// Config retrieves the blockchain's chain configuration.
+	// PluginConfig() ChainConfigurator
+
+	// CurrentHeader retrieves the current header from the local chain.
+	CurrentHeader() *types.Header
+
+	// GetHeader retrieves a block header from the database by hash and number.
+	GetHeader(hash core.Hash, number uint64) *types.Header
+
+	// GetHeaderByNumber retrieves a block header from the database by number.
+	GetHeaderByNumber(number uint64) *types.Header
+
+	// GetHeaderByHash retrieves a block header from the database by its hash.
+	GetHeaderByHash(hash core.Hash) *types.Header
+
+	// GetTd retrieves the total difficulty from the database by hash and number.
+	GetTd(hash core.Hash, number uint64) *big.Int
+}
+
+// ChainReader defines a small collection of methods needed to access the local
+// blockchain during header and/or uncle verification.
+type ChainReader interface {
+	ChainHeaderReader
+
+	// GetBlock retrieves a block from the database by hash and number.
+	GetBlock(hash core.Hash, number uint64) *types.Block
+}
+
+type Configurator interface {
+	ChainConfigurator
+	GenesisBlocker
+}
+
+type ChainConfigurator interface {
+	String() string
+
+	ProtocolSpecifier
+	Forker
+	ConsensusEnginator // Consensus Engine
+	// CHTer
+}
+
+type Forker interface {
+	// IsEnabled tells if interface has met or exceeded a fork block number.
+	// eg. IsEnabled(c.GetEIP1108Transition, big.NewInt(42)))
+	IsEnabled(fn func() *uint64, n *big.Int) bool
+	IsEnabledByTime(fn func() *uint64, n *uint64) bool
+
+	// ForkCanonHash yields arbitrary number/hash pairs.
+	// This is an abstraction derived from the original EIP150 implementation.
+	GetForkCanonHash(n uint64) core.Hash
+	SetForkCanonHash(n uint64, h core.Hash) error
+	GetForkCanonHashes() map[uint64]core.Hash
+}
+
+type ConsensusEnginator interface {
+	GetConsensusEngineType() ConsensusEngineT
+	MustSetConsensusEngineType(t ConsensusEngineT) error
+	GetIsDevMode() bool
+	SetDevMode(devMode bool) error
+
+	EthashConfigurator
+	CliqueConfigurator
```

Comment:
I feel like we're probably not doing anything with this or the Lyra2Configurator.